controller_manager/test/test_control_loop.cpp

```diff
@@ -0,0 +1,194 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "controller_manager/control_loop.hpp"
+
+#include <chrono>
+#include <numeric>
+#include <thread>
+#include <vector>
+
+#include <gtest/gtest.h>
+
+#include "rclcpp/clock.hpp"
+#include "rclcpp/time.hpp"
+
+using namespace std::literals;
+
+namespace
+{
+rclcpp::Time now()
+{
+  thread_local auto clock = rclcpp::Clock{};
+  return clock.now();
+}
+
+void sleep_until(rclcpp::Time time)
+{
+  std::this_thread::sleep_until(
+    std::chrono::system_clock::time_point(std::chrono::nanoseconds(time.nanoseconds())));
+}
+
+}  // namespace
+
+TEST(TestControlLoop, ExitsWhenNotOk)
+{
+  auto work_count = 0;
+  auto do_work = [&work_count](rclcpp::Time, rclcpp::Duration) { work_count++; };
+
+  // GIVEN ok function that returns false
+  auto ok = []() { return false; };
+
+  // WHEN we run the ControlLoop
+  ControlLoop(1000ns, now, ok, do_work, sleep_until);
+
+  // THEN we expect do_work was never called, so the work_count should be 0
+  EXPECT_EQ(work_count, 0);
+}
+
+TEST(TestControlLoop, DurationIsPeriod)
+{
+  auto loop_iterations = 0;
+  auto ok = [&loop_iterations]() {
+    loop_iterations++;
+    return loop_iterations < 10;
+  };
+
+  // GIVEN do_work function that records duration values
+  std::vector<rclcpp::Duration> durations;
+  auto do_work = [&durations](rclcpp::Time, rclcpp::Duration duration) {
+    durations.push_back(duration);
+  };
+
+  // WHEN we run the ControlLoop
+  ControlLoop(10'000'000ns, now, ok, do_work, sleep_until);
+
+  // THEN we expect the average duration to be close to the period
+  // accurate to 0.1ms. The period we set was 10ms.
+  auto const average_duration = std::chrono::nanoseconds{
+    std::reduce(durations.cbegin() + 1, durations.cend(), rclcpp::Duration{0, 0}).nanoseconds() /
+    (durations.size() - 1)};
+  EXPECT_NEAR(average_duration.count(), 10'000'000, 100'000);
+}
+
+TEST(TestControlLoop, FirstDurationIsSmall)
+{
+  auto loop_iterations = 0;
+  auto ok = [&loop_iterations]() {
+    loop_iterations++;
+    return loop_iterations < 2;
+  };
+
+  // GIVEN do_work function that records duration values
+  std::vector<rclcpp::Duration> durations;
+  auto do_work = [&durations](rclcpp::Time, rclcpp::Duration duration) {
+    durations.push_back(duration);
+  };
+
+  // WHEN we run the ControlLoop
+  ControlLoop(10'000'000ns, now, ok, do_work, sleep_until);
+
+  // THEN we expect the first duration to be near 0
+  EXPECT_NEAR(durations.at(0).nanoseconds(), 0, 1'000);
+}
+
+TEST(TestControlLoop, DoWorkDelayInDuration)
```

Comment:
```suggestion
TEST(TestControlLoop, DoWorkTimeGreaterThanPeriod)
```

nit - I think this would be clearer